multicluster/service-mirror/cluster_watcher.go

```diff
@@ -815,3 +906,331 @@ func (rcsw *RemoteClusterServiceWatcher) createOrUpdateEndpoints(ctx context.Con
+			if addr.TargetRef == nil {
+				// ignore addresses w/o a backing pod (if they exist)
+				continue
+			}
+
+			refLocalName := rcsw.mirroredResourceName(addr.TargetRef.Name)
+			createdService, err := rcsw.createHeadlessMirroredHost(ctx, addr.TargetRef, remoteService, refLocalName)
+			if err != nil {
+				rcsw.log.Errorf("failed to create headless mirrored hosts for %s: %v", refLocalName, err)
+				continue
+			}
+
+			hosts[addr.TargetRef.Name] = struct{}{}
+			newAddresses = append(newAddresses, corev1.EndpointAddress{
+				Hostname: addr.TargetRef.Name,
+				IP:       createdService.Spec.ClusterIP,
+			})
+
+		}
+
+		if len(newAddresses) == 0 {
+			continue
+		}
+
+		subsetsToCreate = append(subsetsToCreate, corev1.EndpointSubset{
+			Addresses: newAddresses,
+			Ports:     subset.DeepCopy().Ports,
+		})
+	}
+
+	endpointsToCreate := &corev1.Endpoints{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      localServiceName,
+			Namespace: remoteService.Namespace,
+			Labels: map[string]string{
+				consts.MirroredResourceLabel:  "true",
+				consts.RemoteClusterNameLabel: rcsw.link.TargetClusterName,
+			},
+			Annotations: map[string]string{
+				consts.RemoteServiceFqName: fmt.Sprintf("%s.%s.svc.%s", remoteService.Name, remoteService.Namespace, rcsw.link.TargetClusterDomain),
+			},
+		},
+		Subsets: subsetsToCreate,
+	}
+
+	if rcsw.link.GatewayIdentity != "" {
+		endpointsToCreate.Annotations[consts.RemoteGatewayIdentity] = rcsw.link.GatewayIdentity
+	}
+
+	rcsw.log.Infof("Creating a new root Endpoints for %s", serviceInfo)
+	if _, err := rcsw.localAPIClient.Client.CoreV1().Endpoints(remoteService.Namespace).Create(ctx, endpointsToCreate, metav1.CreateOptions{}); err != nil {
+		// we clean up after ourselves
+		rcsw.localAPIClient.Client.CoreV1().Services(remoteService.Namespace).Delete(ctx, localServiceName, metav1.DeleteOptions{})
+		// and retry
+		return RetryableError{[]error{err}}
+	}
+
+	return nil
+}
+
+func (rcsw *RemoteClusterServiceWatcher) deleteHeadlessMirrors(namespace, name string) error {
+	ep, err := rcsw.localAPIClient.Endpoint().Lister().Endpoints(namespace).Get(name)
+	if err != nil {
+		return err
+	}
+
+	if err = checkEndpointsValid(ep); err != nil {
+		return err
+	}
+
+	for _, subset := range ep.Subsets {
+		for _, address := range subset.Addresses {
+			if address.Hostname == "" {
+				continue
+			}
+			rcsw.eventsQueue.Add(&RemoteServiceDeleted{
+				Name:      address.Hostname,
+				Namespace: namespace,
+			})
+		}
+	}
+	return nil
+}
+
+func (rcsw *RemoteClusterServiceWatcher) createHeadlessMirroredHost(ctx context.Context, targetRef *corev1.ObjectReference, remoteService *corev1.Service, refLocalName string) (*corev1.Service, error) {
+	gatewayAddresses, err := rcsw.resolveGatewayAddress()
+	if err != nil {
+		return nil, err
+	}
+
+	hostAnnotations := map[string]string{
+		// needed to detect real changes
+		consts.RemoteResourceVersionAnnotation: targetRef.ResourceVersion,
```

Comment:
Good question. I was hesitant to give all of the Endpoint Mirrors the same (shared) resource version. I think we use it to detect change only in `createOrUpdateService`, but I also thought it would make more sense to have each Endpoint Mirror share the resource version with its respective pod in the target cluster.

Throughout the PR, I chose to ignore Endpoints addresses without a backing TargetRef -- I thought we wouldn't want to mirror a synthetic Endpoints object. I'm interested to hear if you think we should account for those as well.